### PR TITLE
Leave `serde` as the only serialization method

### LINKED ISSUE #95: Re-evaluate encoding choice for human-readable formats.
This issue also concerns `nucypher-core`.

Currently, when serializing to human-readable formats with `serde`, we are using hex encoding for anything under 33 bytes (that is, `PublicKey` in `umbral-pre`, and `Address`, `HRAC` and `FleetStateChecksum` in `nucypher-core`), and base64 for the rest. Is this a good choice? Other options:

- Use hex for everything, like, say, Ethereum RPC does. This will make the code simpler, but it will increase the size of serialized objects. Most of our objects are rather small, of the order of hundreds of bytes; the biggest offenders will be `EncryptedTreasureMap` and `MessageKit`, whose `ciphertext` fields can be large. 
- Use https://github.com/multiformats/multibase , and keep the current encoding scheme. This is more future-proof, and provides one entry point for encoding/decoding, but may be confusing for users.

### LINKED ISSUE #109: Use `serde` exclusively for serialization
Currently we have a two-step system: `umbral` objects are serialized manually as byte blobs (with `SerializableToArray` trait) and then the result can be (optionally) serialized with `serde` (to enable composition in `nucypher-core`). Originally, this choice was made for the following reasons:
- compatibility with `ByteStringSplitter` (not important anymore)
- support for no-alloc targets (`serde` can handle that, for certain formats)
- efficient byte array serialization (now I know how to do it with `serde`)

I don't think there was anything else, was there?

So I think it might be the time to retire `SerializableToArray`/`DeserializableFromArray` and just use `serde`. The upside is significant code simplification and less boilerplate when adding new structures (which I bumped into when working on #107). The downsides are:
- Changed ABI, so protocol versions in `nucypher-core` will have to be bumped.
- Slightly less efficient packaging in binary formats, since `serde` does not support constant-sized arrays, so every byte blob (curve point, hash, or signature) will have its length serialized too.
- Less efficient packaging in text formats, since we will only be able to mandate encoding for "leaf" types (aforementioned points, hashes, and signatures), so instead of encoding, say, the whole cfrag in base64 as a byte blob, we will have a hierarchical structure with hex-encoded components. (Porter can still encode cfrags in binary and then in base64 if it prefers)
- All the users who need serialization (and `umbral` is not particularly useful without it) will have to use `serde` and a crate for whatever format they use, which are somewhat heavy dependencies. 

Speaking of the last point, we might as well just make `serde` a mandatory dependency (this will help avoid some problems with documenting the `serde` trait support, too).

Also this will resolve #95 - we will just use hex for small objects and that's it.

@piotr-roslaniec, @jMyles, @KPrasch, @theref - thoughts?

### PR DESCRIPTION
**Pull request by fjarri**
Fixes #109 
Fixes #95 (or at least invalidates it)

See https://github.com/nucypher/nucypher-core/pull/41 for the preview of how it will affect `nucypher-core`

Changes:
- Remove `HasTypeName` trait, and use `any::type_name()` where some name is still required.
- Add `serde::Serialize`/`Deserialize` implementations for `Caspule`, `(Verified)CapsuleFrag`, `(Verified)KeyFrag`, `PublicKey`, and `Signature`.
- Remove `VerifiedCapsuleFrag::from_verified_bytes()` and `VerifiedKeyFrag::from_verified_bytes()`. For this behavior, deserialize into `CapsuleFrag` or `KeyFrag` and call `skip_verification()`.
- Remove `DeserializableFromArray`, `RepresentableAsArray`, `SerializableToArray`, `SerializableToSecretArray`. 
- Add `DefaultSerialize`/`DefaultDeserialize` for Umbral objects (`Caspule`, `(Verified)CapsuleFrag`, `(Verified)KeyFrag`) which uses MessagePack via `serde` (gated by `default-serialization` feature). This is what the serialization methods in the bindings use.
- Add `to_compressed_bytes()`/`try_from_compressed_bytes()` for `PublicKey` to give access to non-serde representation (just 33 bytes). These are exposed in the bindings in lieu of the "default" methods (e.g. `__bytes__()` or `toBytes()`).
- Add `to_der_bytes()`/`try_from_der_bytes()` for `Signature` to give access to non-serde representation. These are exposed in the bindings in lieu of the "default" methods (e.g. `__bytes__()` or `toBytes()`).
- Add `SecretKeyFactory::make_secret()` to provide a more straightforward replacement of `SecretKeyFactory::make_secret_key()`->`SecretKey::to_secret_bytes()`.
- `Capsule` no longer implements `Copy` (it's ~128 bytes, I don't think it's wise to have it)

Internal changes:
- Add `Serialize`/`Deserialize` implementations for `CurveScalar` and `CurvePoint`.
- Hash in Python bindings is calculated just with SHA2(byte_representation), instead of bothering GIL and Python builtins.

Remains to be decided:
- `serde_bytes` module may be better extracted into its own crate or added to some existing bytes-handling crate for `serde`

### CODE REVIEW

**Comment by theref** [comment]
what does `der` in `to_der_bytes()`/`try_from_der_bytes()` mean?

**Comment by fjarri** [comment]
It's the standard format for serializing signatures or certificates, see https://en.wikipedia.org/wiki/X.690#DER_encoding

File: umbral-pre/src/bindings_python.rs [hunk 1]
```diff
@@ -11,82 +11,68 @@
...
 
-fn hash<T, U>(obj: &T) -> PyResult<isize>
-where
-    T: AsRef<U>,
-    U: SerializableToArray + HasTypeName,
-{
-    let serialized = obj.as_ref().to_array();
+fn type_name<U>() -> &'static str {
+    // TODO: for a slightly better user experience we can remove qualifiers here,
+    // because the returned string will be something like "crate_name::module_name::TypeName"
```
Comment (piotr-roslaniec) [inline umbral-pre/src/bindings_python.rs:53]: I personally have nothing against fully-qualified names in this case.

**Comment by fjarri** [comment]
Yes, there is a bunch - see https://github.com/serde-rs/bytes/issues/37#issuecomment-1373926255 . `serde-bytes-repr` in particular requires one to write boilerplate, which I would like to avoid.

**Pull request merged**